libs/core/uagents_core/utils/registration.py

```diff
@@ -0,0 +1,229 @@
+import hashlib
+import json
+
+import requests
+from typing import Optional, Union, List, Dict
+from pydantic import BaseModel, Field
+import time
+import urllib.parse
+
+from uagents_core.crypto import Identity
+from uagents_core.types import AgentEndpoint, AgentType
+from uagents_core.config import (
+    DEFAULT_AGENTVERSE_URL,
+    DEFAULT_ALMANAC_API_PATH,
+    DEFAULT_MAILBOX_API_PATH,
+    DEFAULT_CHALLENGE_PATH,
+)
+from uagents_core.logger import get_logger
+
+
+
+
+logger = get_logger("uagents_core.utils.registration")
+
+
+class VerifiableModel(BaseModel):
+    agent_address: str
+    signature: Optional[str] = None
+    timestamp: Optional[int] = None
+
+    def sign(self, identity: Identity):
+        self.timestamp = int(time.time())
+        digest = self._build_digest()
+        self.signature = identity.sign_digest(digest)
+
+    def verify(self) -> bool:
+        return self.signature is not None and Identity.verify_digest(
+            self.agent_address, self._build_digest(), self.signature
+        )
+
+    def _build_digest(self) -> bytes:
+        sha256 = hashlib.sha256()
+        sha256.update(
+            json.dumps(
+                self.model_dump(exclude={"signature"}),
+                sort_keys=True,
+                separators=(",", ":"),
+            ).encode("utf-8")
+        )
+        return sha256.digest()
+
+
+class AgentRegistrationAttestation(VerifiableModel):
+    protocols: List[str]
+    endpoints: List[AgentEndpoint]
+    metadata: Optional[Dict[str, Union[str, Dict[str, str]]]] = None
+
+
+class RegistrationRequest(BaseModel):
+    address: str
+    challenge: str
+    challenge_response: str
+    agent_type: AgentType
+    endpoint: Optional[str] = None
+
+
+class RegistrationResponse(BaseModel):
+    success: bool
+
+
+class ChallengeRequest(BaseModel):
+    address: str
+
+
+class ChallengeResponse(BaseModel):
+    challenge: str
+
+
+class AgentUpdates(BaseModel):
+    name: str = Field(min_length=1, max_length=80)
+    readme: Optional[str] = Field(default=None, max_length=80000)
+    avatar_url: Optional[str] = Field(default=None, max_length=4000)
+    agent_type: Optional[AgentType] = "custom"
+
+
+def register_with_agentverse(
+    identity: Identity,
+    url: str,
+    agentverse_token: str,
+    agent_title: str,
+    readme: str,
+    *,
+    protocol_digest: str,
+    agentverse_url: Optional[str] = None,
+    agent_type: AgentType = "custom",
+):
```

Comment:
Can we at least use the same API that we use in the async version of this: https://github.com/fetchai/uAgents/blob/main/python/src/uagents/mailbox.py#L87-L92?